Keep schema registry up to date

**Problem description**
Schema Registry will upgrade the internal Avro to 1.11 which will break the compatibility of the current implementation.

**Problem location**
Avro deserialization schema

**Suggestions for an improvement**

To add some more details, the schema registry update would break the tests of interpreting with Avro's `GenericRecord` in both reader and writer, more specificly, this test case for example: https://github.com/pravega/flink-connectors/blob/336172a3f955adb209e673dd45240dfd74eb45a9/src/test/java/io/pravega/connectors/flink/FlinkPravegaSchemaRegistryReaderTestITCase.java#L85

Error stacktrace:
```
2022-04-27 17:27:43,740 ERROR io.pravega.connectors.flink.FlinkPravegaReader               [] - Exception occurred while creating a Pravega EventStreamReader to read events
org.apache.flink.streaming.runtime.tasks.ExceptionInChainedOperatorException: Could not forward element to next operator
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:99) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:57) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:56) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:29) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollect(StreamSourceContexts.java:418) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collect(StreamSourceContexts.java:513) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.StreamSourceContexts$SwitchingOnClose.collect(StreamSourceContexts.java:103) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at io.pravega.connectors.flink.FlinkPravegaReader.emitEvent(FlinkPravegaReader.java:357) ~[main/:?]
    at io.pravega.connectors.flink.FlinkPravegaReader.run(FlinkPravegaReader.java:321) [main/:?]
    at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:116) [flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:73) [flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:323) [flink-streaming-java_2.12-1.14.0.jar:1.14.0]
Caused by: com.esotericsoftware.kryo.KryoException: java.lang.UnsupportedOperationException
Serialization trace:
reserved (org.apache.avro.Schema$Field)
fieldMap (org.apache.avro.Schema$RecordSchema)
schema (org.apache.avro.generic.GenericData$Record)
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:143) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:21) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:657) ~[kryo-2.24.0.jar:?]
    at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.copy(KryoSerializer.java:273) ~[flink-core-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:80) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    ... 12 more
Caused by: java.lang.UnsupportedOperationException
    at java.util.Collections$UnmodifiableCollection.add(Collections.java:1060) ~[?:?]
    at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:109) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:22) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:143) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:21) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528) ~[kryo-2.24.0.jar:?]
    at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:657) ~[kryo-2.24.0.jar:?]
    at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.copy(KryoSerializer.java:273) ~[flink-core-1.14.0.jar:1.14.0]
    at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:80) ~[flink-streaming-java_2.12-1.14.0.jar:1.14.0]
    ... 12 more
```

We first try to offer an higher-level `withDeserializationSchemaFromRegistry` interface in #422 with much detail encapsulated to handle all the json, avro and protobuf format. After the avro is passively upgraded, we find that the way we handle the produced type for `GenericRecord` is not correct. It should contain the schema information as `new GenericRecordAvroTypeInfo(SCHEMA)` for Flink to deserialize, but we just pass this `TypeInformation.of(GenericRecord.class)` which causes this error under Avro 1.11.

Now we would decide to remove this `withDeserializationSchemaFromRegistry` API, and just keep the `DeserializerFromSchemaRegistry` and let the users to wrap this freely with the correct type information.

--- a/gradle.properties
+++ b/gradle.properties
@@ -37,8 +37,8 @@ nettyVersion=4.1.65.Final
 
 # Version and base tags can be overridden at build time.
 connectorVersion=0.11.0-SNAPSHOT
-pravegaVersion=0.11.0-3038.fc1f5c3-SNAPSHOT
-schemaRegistryVersion=0.4.0-78.f1b3eef-SNAPSHOT
+pravegaVersion=0.11.0
+schemaRegistryVersion=0.4.0-80.1629800-SNAPSHOT
 apacheCommonsVersion=3.7
 
 # These properties are only needed for publishing to maven central

--- a/src/main/java/io/pravega/connectors/flink/FlinkPravegaReader.java
+++ b/src/main/java/io/pravega/connectors/flink/FlinkPravegaReader.java
@@ -27,8 +27,6 @@ import io.pravega.client.stream.ReaderGroupConfig;
 import io.pravega.client.stream.ReaderGroupNotFoundException;
 import io.pravega.client.stream.Stream;
 import io.pravega.client.stream.TruncatedDataException;
-import io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry;
-import io.pravega.connectors.flink.serialization.PravegaDeserializationSchema;
 import io.pravega.connectors.flink.serialization.PravegaDeserializationSchemaWithMetadata;
 import io.pravega.connectors.flink.watermark.AssignerWithTimeWindows;
 import org.apache.flink.api.common.ExecutionConfig;
@@ -717,20 +715,6 @@ public class FlinkPravegaReader<T>
          */
         public Builder<T> withDeserializationSchema(DeserializationSchema<T> deserializationSchema) {
             this.deserializationSchema = deserializationSchema;
-            return builder();
-        }
-
-        /**
-         * Sets the deserialization schema from schema registry. It supports Json, Avro and Protobuf format.
-         *
-         * @param groupId The group id in schema registry
-         * @param tClass  The class describing the deserialized type.
-         * @return Builder instance.
-         */
-        @SuppressWarnings("unchecked")
-        public Builder<T> withDeserializationSchemaFromRegistry(String groupId, Class<T> tClass) {
-            this.deserializationSchema = new PravegaDeserializationSchema<>(tClass,
-                    new DeserializerFromSchemaRegistry<>(getPravegaConfig(), groupId, tClass));
             return builder();
         }
 

--- a/src/main/java/io/pravega/connectors/flink/serialization/PravegaDeserializationSchema.java
+++ b/src/main/java/io/pravega/connectors/flink/serialization/PravegaDeserializationSchema.java
@@ -67,10 +67,10 @@ public class PravegaDeserializationSchema<T>
             this.typeInfo = TypeInformation.of(typeClass);
         } catch (InvalidTypesException e) {
             throw new IllegalArgumentException(
-                    "Due to Java's type erasure, the generic type information cannot be properly inferred. " + 
-                    "Please pass a 'TypeHint' instead of a class to describe the type. " +
-                    "For example, to describe 'Tuple2<String, String>' as a generic type, use " +
-                    "'new PravegaDeserializationSchema<>(new TypeHint<Tuple2<String, String>>(){}, serializer);'"
+                    "Due to Java's type erasure, the generic type information cannot be properly inferred. " +
+                            "Please pass a 'TypeHint' instead of a class to describe the type. " +
+                            "For example, to describe 'Tuple2<String, String>' as a generic type, use " +
+                            "'new PravegaDeserializationSchema<>(new TypeHint<Tuple2<String, String>>(){}, serializer);'"
             );
         }
     }

--- a/src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java
+++ b/src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java
@@ -35,6 +35,8 @@ import io.pravega.client.stream.TruncatedDataException;
 import io.pravega.client.stream.impl.EventPointerImpl;
 import io.pravega.client.stream.impl.EventReadImpl;
 import io.pravega.client.stream.impl.StreamCutImpl;
+import io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry;
+import io.pravega.connectors.flink.serialization.PravegaDeserializationSchema;
 import io.pravega.connectors.flink.serialization.PravegaDeserializationSchemaWithMetadata;
 import io.pravega.connectors.flink.utils.IntegerDeserializationSchema;
 import io.pravega.connectors.flink.utils.IntegerSerializer;
@@ -352,7 +354,8 @@ public class FlinkPravegaReaderTest {
             FlinkPravegaReader.<Integer>builder()
                     .withPravegaConfig(pravegaConfig)
                     .forStream("stream")
-                    .withDeserializationSchemaFromRegistry("stream", Integer.class)
+                    .withDeserializationSchema(new PravegaDeserializationSchema<>(Integer.class,
+                            new DeserializerFromSchemaRegistry<>(pravegaConfig, "stream", Integer.class)))
                     .build();
             fail();
         } catch (NullPointerException e) {
@@ -364,7 +367,8 @@ public class FlinkPravegaReaderTest {
             FlinkPravegaReader.<Integer>builder()
                     .withPravegaConfig(pravegaConfig)
                     .forStream("stream")
-                    .withDeserializationSchemaFromRegistry("stream", Integer.class)
+                    .withDeserializationSchema(new PravegaDeserializationSchema<>(Integer.class,
+                            new DeserializerFromSchemaRegistry<>(pravegaConfig, "stream", Integer.class)))
                     .build();
             fail();
         } catch (NullPointerException e) {

--- a/src/test/java/io/pravega/connectors/flink/FlinkPravegaSchemaRegistryReaderTestITCase.java
+++ b/src/test/java/io/pravega/connectors/flink/FlinkPravegaSchemaRegistryReaderTestITCase.java
@@ -17,6 +17,8 @@
 package io.pravega.connectors.flink;
 
 import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry;
+import io.pravega.connectors.flink.serialization.PravegaDeserializationSchema;
 import io.pravega.connectors.flink.utils.SchemaRegistryUtils;
 import io.pravega.connectors.flink.utils.SetupUtils;
 import io.pravega.connectors.flink.utils.SuccessException;
@@ -86,12 +88,15 @@ public class FlinkPravegaSchemaRegistryReaderTestITCase {
         final String streamName = RandomStringUtils.randomAlphabetic(20);
         prepareAvroStream(streamName, AvroSchema.of(SCHEMA));
 
+        final PravegaConfig pravegaConfig = SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri());
         final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
         FlinkPravegaReader<GenericRecord> reader = FlinkPravegaReader.<GenericRecord>builder()
                 .forStream(streamName)
                 .enableMetrics(false)
-                .withPravegaConfig(SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri()))
-                .withDeserializationSchemaFromRegistry(streamName, GenericRecord.class)
+                .withPravegaConfig(pravegaConfig)
+                .withDeserializationSchema(new PravegaDeserializationSchema<>(
+                        new GenericRecordAvroTypeInfo(SCHEMA),
+                        new DeserializerFromSchemaRegistry<>(pravegaConfig, streamName, GenericRecord.class)))
                 .build();
 
         env.addSource(reader).addSink(new SinkFunction<GenericRecord>() {
@@ -118,12 +123,14 @@ public class FlinkPravegaSchemaRegistryReaderTestITCase {
         final String streamName = RandomStringUtils.randomAlphabetic(20);
         prepareAvroStream(streamName, AvroSchema.of(SCHEMA));
 
+        final PravegaConfig pravegaConfig = SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri());
         final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
         FlinkPravegaReader<User> reader = FlinkPravegaReader.<User>builder()
                 .forStream(streamName)
                 .enableMetrics(false)
-                .withPravegaConfig(SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri()))
-                .withDeserializationSchemaFromRegistry(streamName, User.class)
+                .withPravegaConfig(pravegaConfig)
+                .withDeserializationSchema(new PravegaDeserializationSchema<>(User.class,
+                        new DeserializerFromSchemaRegistry<>(pravegaConfig, streamName, User.class)))
                 .build();
 
         env.addSource(reader).addSink(new SinkFunction<User>() {
@@ -150,12 +157,14 @@ public class FlinkPravegaSchemaRegistryReaderTestITCase {
         final String streamName = RandomStringUtils.randomAlphabetic(20);
         prepareJsonStream(streamName, JSONSchema.of(MyTest.class));
 
+        final PravegaConfig pravegaConfig = SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri());
         final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
         FlinkPravegaReader<MyTest> reader = FlinkPravegaReader.<MyTest>builder()
                 .forStream(streamName)
                 .enableMetrics(false)
-                .withPravegaConfig(SETUP_UTILS.getPravegaConfig().withSchemaRegistryURI(SCHEMA_REGISTRY_UTILS.getSchemaRegistryUri()))
-                .withDeserializationSchemaFromRegistry(streamName, MyTest.class)
+                .withPravegaConfig(pravegaConfig)
+                .withDeserializationSchema(new PravegaDeserializationSchema<>(MyTest.class,
+                        new DeserializerFromSchemaRegistry<>(pravegaConfig, streamName, MyTest.class)))
                 .build();
 
         env.addSource(reader).addSink(new SinkFunction<MyTest>() {

--- a/src/test/java/io/pravega/connectors/flink/FlinkPravegaSchemaRegistryWriterTestITCase.java
+++ b/src/test/java/io/pravega/connectors/flink/FlinkPravegaSchemaRegistryWriterTestITCase.java
@@ -34,6 +34,7 @@ import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.generic.GenericRecordBuilder;
 import org.apache.commons.lang3.RandomStringUtils;
 import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.formats.avro.typeutils.GenericRecordAvroTypeInfo;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.streaming.api.functions.source.SourceFunction;
 import org.junit.AfterClass;
@@ -131,7 +132,7 @@ public class FlinkPravegaSchemaRegistryWriterTestITCase {
             @Override
             public void cancel() {
             }
-        }).addSink(writer);
+        }, new GenericRecordAvroTypeInfo(SCHEMA)).addSink(writer);
 
         try {
             env.execute("Schema Registry Read");

